File: app/src/main/java/com/lucasurbas/counter/ui/explore/ExploreAdapter.java
package com.lucasurbas.counter.ui.explore;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import com.hannesdorfmann.adapterdelegates3.AdapterDelegatesManager;
import com.lucasurbas.counter.ui.explore.model.UiListItem;
import com.lucasurbas.counter.utils.ListItemDiffCallback;

import java.util.ArrayList;
import java.util.List;

public class ExploreAdapter extends RecyclerView.Adapter {

    private static final int TYPE_HEADER = 1;
    private static final int TYPE_COUNTER = 2;

    private AdapterDelegatesManager<List<UiListItem>> delegatesManager;
    private List<UiListItem> itemList = new ArrayList<>();

    public ExploreAdapter(CountertemAdapterDelegate.OnCounterItemClickListener counterItemClickListener) {
        delegatesManager = new AdapterDelegatesManager<>();
        delegatesManager.addDelegate(TYPE_HEADER, new HeaderItemAdapterDelegate())
                .addDelegate(TYPE_COUNTER, new CountertemAdapterDelegate(counterItemClickListener));
    }

    @Override
    public int getItemViewType(int position) {
        return delegatesManager.getItemViewType(itemList, position);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return delegatesManager.onCreateViewHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        delegatesManager.onBindViewHolder(itemList, position, holder);
    }

    @Override
    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
        delegatesManager.onViewRecycled(holder);
    }

    @Override
    public void onViewDetachedFromWindow(@NonNull RecyclerView.ViewHolder holder) {
        delegatesManager.onViewDetachedFromWindow(holder);
    }

    @Override
    public int getItemCount() {
        return itemList.size();
    }

    @Override
    public long getItemId(int position) {
        return itemList.get(position).getId();
    }

    public int getSpanSize(int position, int spanCount) {
        switch (getItemViewType(position)) {
            case TYPE_HEADER:
                return spanCount;
            default:
                return 1;
        }
    }

    public void updateItemList(@Nullable List<UiListItem> newItemList) {
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new ListItemDiffCallback(itemList, newItemList));
        itemList.clear();
        itemList.addAll(newItemList);
        diffResult.dispatchUpdatesTo(this);
    }
}

File: app/src/main/java/com/lucasurbas/counter/ui/explore/ExploreAdapter.kt
package com.lucasurbas.counter.ui.explore


import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import com.hannesdorfmann.adapterdelegates3.AdapterDelegatesManager
import com.lucasurbas.counter.ui.explore.model.UiListItem
import com.lucasurbas.counter.utils.ListItemDiffCallback
import java.util.*

class ExploreAdapter(
        counterItemClickListener: (Int, View) -> Unit
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private val TYPE_HEADER = 1
        private val TYPE_COUNTER = 2
    }

    private val delegatesManager: AdapterDelegatesManager<List<UiListItem>>
    private val itemList = ArrayList<UiListItem>()

    init {
        delegatesManager = AdapterDelegatesManager()
        delegatesManager
                .addDelegate(TYPE_HEADER, HeaderItemAdapterDelegate())
                .addDelegate(TYPE_COUNTER, CountertemAdapterDelegate(counterItemClickListener))
    }

    override fun getItemViewType(position: Int): Int {
        return delegatesManager.getItemViewType(itemList, position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return delegatesManager.onCreateViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        delegatesManager.onBindViewHolder(itemList, position, holder)
    }

    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        delegatesManager.onViewRecycled(holder)
    }

    override fun onViewDetachedFromWindow(holder: RecyclerView.ViewHolder) {
        delegatesManager.onViewDetachedFromWindow(holder)
    }

    override fun getItemCount(): Int {
        return itemList.size
    }

    override fun getItemId(position: Int): Long {
        return itemList[position].id.toLong()
    }

    fun getSpanSize(position: Int, spanCount: Int): Int {
        return when (getItemViewType(position)) {
            TYPE_HEADER -> spanCount
            else -> 1
        }
    }

    fun updateItemList(newItemList: List<UiListItem>) {
        val diffResult = DiffUtil.calculateDiff(ListItemDiffCallback(itemList, newItemList))
        itemList.clear()
        itemList.addAll(newItemList)
        diffResult.dispatchUpdatesTo(this)
    }
}
